[Wallet] WeChat Pay

Mayisha opened the pull request:
Fixes #3130 

## Changes proposed in this Pull Request:
This PR introduces the support for Wallet payments with the "New checkout experience" feature enabled and implements the WeChat Pay payment method.

## Testing instructions
- Checkout to this branch `task/3130-wechat-payment`.
- Connect to a Stripe account that is supported country and set store currency to a supported currency for WeChat Pay. ([supported currency and country for WeChat](https://docs.stripe.com/payments/wechat-pay/accept-a-payment?ui=direct-api#supported-currencies))
- Configure the webhook for your site.

**Admin Settings**
- As a merchant, go to the Stripe extension settings page `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods`
- Confirm that you can see WeChat Pay in the list of payment methods.

![wechat settings](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/334a17ed-5d39-495f-aaf1-728a26132388)

- Change the store currency to an unsupported currency for WeChat and confirm that there is a currency requirement pill beside the payment method.

**Checkout**
- As a shopper add a product to your cart and go to the checkout page.
- Confirm that you can see WeChat Pay as an option.
- Click the place order button and confirm that the WeChat Pay modal is displayed.
- Close the modal without paying. You should still be on the checkout page.

![wechat checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/9bc81e88-aeba-47ab-9f56-3c9c827d5b40)

- Click the place order button again and simulate a successful payment via the `simulate scan` button.
- Check the relevant order in your `wp-admin` and confirm that the order status is set to `processing`.
- Now as a shopper, add another product to your cart and go to the checkout page.
- Click the place order button again and simulate a failed payment via the `simulate scan` button.
- Check the relevant order in your `wp-admin` and confirm that the order status is set to `failed`.

**Refund**
- Select any of the completed orders in the Stripe dashboard and refund it.
- Go to your store and confirm that the order is now refunded

Mayisha committed:
create upe class for wechat

Mayisha committed:
show wechat in the payment method list of the settings page

Mayisha committed:
Specify the 'client' in the 'payment_method_options'
- this param is mandatory for wechat pay for create intend request

Mayisha committed:
add icon

Mayisha committed:
show wechat pay modal when place order button is clicked on the checkout page

Mayisha committed:
show wechat pay on block checkout

Mayisha committed:
fix webhook to set the correct order status after wechat payment

Mayisha committed:
Merge branch 'develop' into task/3130-wechat-payment

Mayisha committed:
add changelog entry

james-allan commented on client/classic/upe/payment-processing.js:
```diff
@@ -373,3 +373,81 @@ export const confirmVoucherPayment = async ( api, jQueryForm ) => {
+		}
+
+		// Once the customer closes the voucher and there are no errors, redirect them to the order received page.
+		window.location.href = decodeURIComponent( partials[ 4 ] );
```
```suggestion
```

I believe we shouldn't redirect the customer if they close the modal when paying with a wallet. 

We do this redirect for voucher payment methods because there's no further action the customer needs to do on the store once they have copied the voucher. 

If you open the WeChat modal but then don't complete the payment, we should leave the customer on the checkout.

james-allan replied:
It will be good to keep the pay for order redirect param though because I'll need it for Cash App.

Mayisha replied:
Fixed this in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3138/commits/8a1b843dcc54b064f45a95b2651587745993c9ef

james-allan commented on client/classic/upe/deferred-intent.js:
```diff
@@ -104,28 +105,38 @@ jQuery( function ( $ ) {
 	}
 
-	// On every page load and on hash change, check to see whether we should display the Boleto or Oxxo modal.
+	// On every page load and on hash change, check to see whether we should display the Voucher( Boleto/Oxxo) or Wallet (CashApp/WeChat Pay) modal.
```
```suggestion
	// On every page load and on hash change, check to see whether we should display the Voucher (Boleto/Oxxo) or Wallet (CashApp/WeChat Pay) modal.
```

Just a minor grammatical error.

james-allan commented on client/classic/upe/payment-processing.js:
```diff
@@ -373,3 +373,81 @@ export const confirmVoucherPayment = async ( api, jQueryForm ) => {
+ * Handles displaying the CashApp or WeChat modal to the customer and then redirecting
+ * them to the order received page once they authenticate the payment.
+ *
+ * When processing a payment for a wallet payment methods on the checkout or order pay page,
```
```suggestion
 * When processing a payment for a wallet payment method on the checkout or order pay page,
```

james-allan commented on assets/images/wechat.svg:
If you want to use it, Stripe have one here that I pulled from the payment elements. https://js.stripe.com/v3/fingerprinted/img/payment-methods/icon-pm-wechat-pay-f62a5a27f646cb5f596c610475d14444.svg

It's bigger and I don't think it will need to be padded on the top or bottom like you have in the `wechat-pay/index.js`.

james-allan replied:
>  I don't think it will need to be padded on the top or bottom like you have in the wechat-pay/index.js.

I might be wrong about this actually. This icon looks imbalanced and too close on the top and bottom borders.

<img width="113" alt="Screenshot 2024-05-21 at 3 16 55 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/e0356930-d0ca-4ab6-828d-a3d2e1fb344b">

Mayisha replied:
> This icon looks imbalanced and too close on the top and bottom borders.

Thanks James. I am keeping the smaller icon for now. Let's revisit this later if needed.

james-allan reviewed (approved):
@Mayisha I've decided to give this a review given the changes here are necessary to continue progress on Cash App. :) 

- **Settings**
   - [x] 👍 Shown for eligible accounts
   - [ ] ⚠️ Hidden for ineligible accounts
      - It's shown for a Brazilian Stripe account for example, even though that merchant will never be eligible. It's not shown on the checkout though - this is because there's no eligible currency returned from `get_supported_currencies()`. 
   - [x] 👍 Shows currency warning
- **Payments**
   - Note payment isn't completed on the Woo side unless you have webhooks set up. 
   - [x] 👍 Classic Checkout payment
       - Noting that the icon is missing on the classic checkout. Changes similar to [these](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3140/files#diff-ff9f40fcdc185cf88b584288b5b59cb3f733a3fd3d53a3ce383e43da20104992R355) will be needed. 
   - [x] 👍 Block Checkout payment
   - [x] 👍 Pay for Order page payment
- **Failed payments**   
   - [x] 👍 All failed payments behave the same. The customer is redirected to the order receive page with a note about it failing, the order has a failed status. 
- **Refunds**   
   - [x] 👍 Full refund
   - [x] 👍 Partial refund
- **Subscriptions**
   - [x] 👍 Not shown when subscriptions are in the cart.

This look good to me. I left some minor feedback. I think there's possibilities to improve the UX for ineligible accounts etc but I'm happy to approve this given the changes here unblock CashApp. I know I've jumped the gun here @Mayisha, but let me know what you think about merging this. :)

Mayisha committed:
fix typo and redirection

Mayisha committed:
add tests

Mayisha commented:
Thanks @james-allan for the review. :chocolate_bar: 

I have added a test and addressed the minor feedback. I am merging this one to unblock CashApp. 

> Hidden for ineligible accounts
It's shown for a Brazilian Stripe account for example, even though that merchant will never be eligible. It's not shown on the checkout though - this is because there's no eligible currency returned from get_supported_currencies().

I am fixing the above issue in a separate PR.
